### ISSUE TITLE
Fallback improvement

### ISSUE THREAD

**Issue by akarienta** [rubenv/angular-gettext#210]
It would be nice to have improved fallback langugage. E.g. I have the string in the locale <code>de</code> but I don't have the string in the locale <code>de_CH</code>. Actually, if the user uses locale <code>de_CH</code>, the string in the default langugae (= English) is used but it is better for user to use the string of locale <code>de</code> instead. Then, if no string for locale <code>de_CH</code> either for locale <code>de</code> exists, the English string is used.

Is it possible? It would be nice to have also a possibility to define this 'fallback waterfall' but the default gettext behaviour is enough I think: <code>(country specific locale, e.g. 'de_CH') -> (general locale, e.g. 'de') -> (the default language = English)</code>.

**Comment by rubenv** [comment]
Yes, we should have this. Never got round to adding it.

Same logic should apply for resolving plural rules.

**Comment by innerpeace0** [comment]
+1

**Comment by rubenv** [comment]
Anyone want to have a go at fixing this?

**Comment by Akarienta** [comment]
Any news about this?

**Comment by crissdev** [comment]
I see @jpzwarte has some implementation in his fork but haven't tested it.

**Comment by rubenv** [comment]
Given that language codes are standardised, it's probably not needed to expose it as an option. The fallback language can be derived automatically: e.g.: `nl_BE` -> `nl`.

**Comment by crissdev** [comment]
@rubenv That's correct but I think there'll still be needed for a general fallback language and use it only if it was defined by the user. Otherwise just use the detected fallback.
To sum, with what @Akarienta said, `de_DE` :point_right: `de` :point_right: `fallbackLanguage`

**Comment by rubenv** [comment]
@crissdev I'm not following you there. Is `fallbackLanguage` something configurable? As opposed to the current behaviour, where we use the (English) source string?

**Comment by Akarienta** [comment]
> Given that language codes are standardised, it's probably not needed to expose it as an option.

Half a truth. There are also some languages which are very similar. E.g. if my mother lang is `sk` I would prefer lang `cs` in the app rather then `en` because I understands it natively without any problem (not the case of English).

**Comment by crissdev** [comment]
I can see what I forgot to add.

`de_DE` :point_right: `de` :point_right: `fallbackLanguage` :point_right: `(English source string)`

**Comment by Akarienta** [comment]
@crissdev This solution would be realy nice.

**Comment by rubenv** [comment]
@crissdev @Akarienta That will only work if plural rules are equal for those languages (which we can expect for variants of the same base language).

I'd rather not support this, for two reasons:

* The goal for angular-gettext is to keep things as simple as possible. A fallback language as described above is a complex corner case not supported by other gettext implementations.
* Missing strings should be the exception, rather than the rule. If not having English anywhere in your app is crucial: don't ship before you update your translations.

**Comment by jpzwarte** [comment]
@rubenv only automatic fallback by doing nl_NL -> nl is insufficient for our needs. We have specific POs which are content related, not only language related. So being able to specify the actual PO used as a fallback is a requirement for us.

**Comment by Akarienta** [comment]
@rubenv Well that's right, obviously. In fact, a user can change the language in app settings in case I am writing about (if he/she likes).

This solution would be enough then: `de_DE` :point_right: `de` :point_right: `(English source string)`.

**Comment by crissdev** [comment]
Having the fallback as above would be fine (and enough) as well.

**Comment by carlos-** [comment]
+1 thanks!
